File: content/01-arrays-and-hashing/elements_majority.py
"""
Problem: Elements Appearing More than N/K times
GeeksforGeeks: https://www.geeksforgeeks.org/given-an-array-of-of-size-n-finds-all-the-elements-that-appear-more-than-nk-times/
"""

def get_majority_elements(arr: list[int], k: int) -> list[int]:
    """
    Finds elements that appear more than N/K times in the array.
    Time Complexity: O(N)
    Space Complexity: O(K)
    """
    n = len(arr)
    if k <= 1:
        return list(set(arr))
        
    # Step 1: Find candidates (at most k-1 candidates)
    candidates = {}
    for num in arr:
        if num in candidates:
            candidates[num] += 1
        elif len(candidates) < k - 1:
            candidates[num] = 1
        else:
            # Decrease count of all candidates
            to_remove = []
            for candidate in candidates:
                candidates[candidate] -= 1
                if candidates[candidate] == 0:
                    to_remove.append(candidate)
            for cand in to_remove:
                del candidates[cand]
                
    # Step 2: Validate candidates
    res = []
    threshold = n // k
    for candidate in candidates:
        count = arr.count(candidate)
        if count > threshold:
            res.append(candidate)
            
    return sorted(res)

if __name__ == "__main__":
    assert get_majority_elements([3, 1, 2, 2, 1, 2, 3, 3], 4) == [2, 3]
    assert get_majority_elements([9, 8, 7, 9, 2, 9, 7], 3) == [9]
    print("Python: get_majority_elements passed tests!")

File: content/01-arrays-and-hashing/elements_majority.js
/**
 * Problem: Elements Appearing More than N/K times
 * GeeksforGeeks: https://www.geeksforgeeks.org/given-an-array-of-of-size-n-finds-all-the-elements-that-appear-more-than-nk-times/
 */

function getMajorityElements(arr, k) {
    /**
     * Finds elements that appear more than N/K times.
     * Time Complexity: O(N)
     * Space Complexity: O(K)
     */
    const n = arr.length;
    if (k <= 1) return Array.from(new Set(arr)).sort((a, b) => a - b);
    
    // Step 1: Find candidates
    let candidates = new Map();
    for (let i = 0; i < n; i++) {
        const num = arr[i];
        if (candidates.has(num)) {
            candidates.set(num, candidates.get(num) + 1);
        } else if (candidates.size < k - 1) {
            candidates.set(num, 1);
        } else {
            const nextCandidates = new Map();
            for (const [key, value] of candidates.entries()) {
                if (value > 1) {
                    nextCandidates.set(key, value - 1);
                }
            }
            candidates = nextCandidates;
        }
    }
    
    // Step 2: Validate candidates
    const res = [];
    const threshold = Math.floor(n / k);
    for (const candidate of candidates.keys()) {
        let count = 0;
        for (let i = 0; i < n; i++) {
            if (arr[i] === candidate) count++;
        }
        if (count > threshold) {
            res.push(candidate);
        }
    }
    
    return res.sort((a, b) => a - b);
}

if (require.main === module) {
    const assert = require('assert');
    assert.deepStrictEqual(getMajorityElements([3, 1, 2, 2, 1, 2, 3, 3], 4), [2, 3]);
    assert.deepStrictEqual(getMajorityElements([9, 8, 7, 9, 2, 9, 7], 3), [9]);
    console.log("JS: getMajorityElements passed tests!");
}
